pandas/core/apply.py

```diff
@@ -358,20 +360,23 @@ def agg_list_like(self) -> DataFrame | Series:
             for index, col in enumerate(selected_obj):
                 colg = obj._gotitem(col, ndim=1, subset=selected_obj.iloc[:, index])
                 try:
-                    new_res = colg.aggregate(arg)
+                    with warnings.catch_warnings(record=True) as w:
+                        new_res = colg.aggregate(arg)
+                        if len(w) > 0:
```

Comment:
are _all_ warnings that happen here going to represent a failure?

Reply:
Woah - thanks. No, probably not. Will inspect the warnings themselves.

Reply:
This got a bit hairy, I wasn't able to find a good way to capture and suppress only our warnings while passing through any other warnings _and_ being able to determine if we raised. Supplying a filter within the catch_warnings context manager means we can't fell if we raised; using record=False there means all user warnings would also be suppressed.

If there are any suggestions for a better implementation, it would be appreciated.

Reply:
what happens if you dont catch warnings at all here?

Reply:
Running

    def foo(x):
       return x.prod()
    df = pd.DataFrame({'a': [1, 2], 'b': ['x', 'y'], 'c': ['z', 'w'], 'd': ['a', 'b']})
    df.agg([foo, lambda x: x.sum()])

gives

    FutureWarning: ['foo'] did not aggregate successfully.
    FutureWarning: ['foo'] did not aggregate successfully.
    FutureWarning: ['foo'] did not aggregate successfully.

Reply:
So wouldn't the thing to do be to find where that is being issues from and extend the warning message there?

Reply:
For a list of aggregators, we loop over the columns of the DataFrame and call agg on each Series. Each of the three warnings above is produced within the separate calls to Series.agg. We can't remove the warning from Series.agg as users may directly be calling it.

Reply:
Is the Series.agg call warning bc e.g. `ser.agg([foo, bar, baz])` is dropping `bar`?  If so, could we just deprecate that too?

Reply:
Yes - that is being deprecated in this PR as well. Each of the three messages I posted in the example above is being emitted because of deprecating for Series.agg (which is called by DataFrame.agg). I had clipped them to avoid a long line, the full message for each of these three is:

> FutureWarning: ['foo'] did not aggregate successfully. If any error is raised this will raise in a future version of pandas. Drop these columns/ops to avoid this warning.

Reply:
then why do we need to catch a warning at all?

Reply:
One call to `DataFrame.agg` could give out many warnings, all saying the same thing. I don't think that is desirable behavior.

Reply:
i think there's something like a `filter="once"` to prevent exactly that (and i think its the default)

Reply:
I see - thanks. Digging into this, the only way I can see why I'm getting multiple warnings is because of https://bugs.python.org/issue29672. I haven't confirmed yet if the code within agg ever hits a separate `catch_warnings`; if it does that will cause repeat messages.

Reply:
I'm still going to verify that the use of another catch_warnings (in a different part of the code) is the cause for the multiple messages I am seeing without the special handling in this PR.

Reply:
I've confirmed it is because of the Python bug mentioned above. The code in the sample hits `pandas.core.indexes.base.Index_with_infer`. Here is a MWE to demonstrate the bug (mirroring the structure of `agg_list_like`):

```
def foo():
    with warnings.catch_warnings():
        warnings.filterwarnings("ignore", ".*the Index constructor", FutureWarning)


def bar(ndim):
    if ndim == 1:
        pass
    else:
        for _ in range(3):
            bar(1)

    warnings.warn('test warning', FutureWarning)
    foo()

bar(2)
```

With the call to `foo` from within `bar`, three warning messages are printed. When the call to `foo` is removed, only one is printed.

With this, are we okay with catching the warnings as was merged @jbrockmendel?

Reply:
im fine with it.  a comment to this effect might be worthwhile